windows/src/engine/keyman32/k32_globals.cpp

```diff
@@ -185,6 +189,7 @@ void Globals_UninitProcess()
   TlsFree(dwTlsIndex);
   dwTlsIndex = TLS_OUT_OF_INDEXES;
   LeaveCriticalSection(&csGlobals);
+  DeleteCriticalSection(&csGlobals);
```

Comment:
This is the most critical change. If we attempt to use `csGlobals` after this we'll run into huge problems, so we have to assume that the DLL has been detached from all threads prior to this being called. This is not necessarily a safe assumption to make!

The problem is, if keyman32.dll is reloaded, it typically reloads at the same address, and so `csGlobals` gets reused. It may be that a safer alternative is to allocate memory for `csGlobals` and let it leak: that will effectively ensure it gets relocated and avoid double-initialization.

DLL detach is messy.

Reply:
Clearly... I don't quite understand what's going on here, even from this.  I've probably got an oversimplified view of things.  So, I'll expound on what I think intuitively here - please address the invalid assumptions I've probably made below:

- `csGlobals` stands for "critical section globals"
- `csGlobals` is allocated in static memory.
  - I say this due to the dereference (`&`) being used to grab a pointer to the variable for the "deletion".
- So, "deleting" `csGlobals` basically means zeroing-out the memory, resetting it to prepare for the next DLL attach?
- The worry then is that b/c race conditions, a thread might pollute the "reset" memory space during shutdown, causing trouble on reload due to unexpectedly altered `csGlobals` state?  (This is the least clear part to me.)

Following from that:
- "safer alternative":  allocate `csGlobals` on the heap instead
- `csGlobals` would then store a pointer
- by letting `csGlobals` leak, any existing references (b/c race conditions) to the old memory location don't cause crashes

Of course, memory leaks aren't ideal.  There ought to be some sort of synchronization/concurrency check on use of `csGlobals` so that we can know when it's safe to delete / reset it.
- since this is mentioned as a problem, I'm assuming there are no concurrency checks in place on `csGlobals`.  (Or "insufficient")
- [ ] Before merging, an issue should be made to fix the ensuing memory leak.

Reply:
Pretty good summary 😉 

I don't know what `DeleteCriticalSection` does internally but it will be removing that cs from Windows' list of critical sections in the process if nothing else.

The concurrency check **is** `csGlobals` -- so when deallocating it there is no other check we can make.

According to the world, `DLL_PROCESS_DETACH` will always be the last event we receive so it's safe to do this deallocation there (e.g. https://docs.microsoft.com/en-us/windows/win32/dlls/dllmain, https://blogs.msdn.microsoft.com/larryosterman/2004/06/10/dll_process_detach-is-the-last-thing-my-dlls-going-to-see-right/)

The concern is primarily that we may still have activity on another thread that is stepping into our DLL's code at the time when we do the deallocation. From my understanding this cannot happen for hooks because a hook is guaranteed to be detached from the process before the library is unloaded by Windows. Unless we have a bug elsewhere! So I *think* it's all safe.

Speaking of leaks, a process detach may happen without a thread detach for all threads. The appropriate response is to free the memory we've allocated for every thread left in the process but we don't currently have a shared list of allocated resources (it's all done in thread-local storage as you'd expect). So this means closing Keyman does leak a few KB in each process. Not enough to matter most of the time. And probably not worth the effort to clean up.

Reply:
> So this means closing Keyman does leak a few KB in each process

I assume that means that the memory we leak is technically within the host process (i.e., program receiving keyboard input)?

Reply:
Yes.

Reply:
> The concurrency check **is** `csGlobals` -- so when deallocating it there is no other check we can make.

Interesting - so no separate object to synchronize on for modifications to its contents/state?

Reply:
Then, how often do we expect Keyman to create leaks within a process?  Does Keyman come in and out of a process with regularity?

Reply:
Critical sections are a synchronization primitive -- they are the object that we use for all other synchronization. They of course need initialization and teardown. Perhaps I'm not understanding your question?

Reply:
> Then, how often do we expect Keyman to create leaks within a process? Does Keyman come in and out of a process with regularity?

Only when a user exits Keyman completely. This should not be normal user behaviour... apart from Explorer and a few system processes, most other processes are started and shut down regularly in any case, so the memory leak will be short lived also.

Reply:
But, none of this around leaks is new -- it's been here for 20+ years so we can probably live with it a little longer...

Reply:
Ah, so I didn't understand the nature of `csGlobals`.  I see the "global**s**", I think `struct`, not primitive.

Reply:
`csGlobals` uses the Windows-style variable name "cs" = critical section, "globals" refers to what it is protecting -- our Keyman global variables (of which there are far too many but topic for another [rainy] day).